class-wp-xml-processor.php

```diff
@@ -88,6 +88,28 @@ public static function stream_tokens( $input_stream, $output_stream, $buffer_siz
 		}
 	}
 
+	/**
+	 * Wipes out the processed XML and appends the next chunk of XML to
+	 * any remaining unprocessed XML.
+	 * 
+	 * @param string $next_chunk XML to append.
+	 */
+	public function append_bytes( string $next_chunk )
+	{
+		$this->get_updated_xml();
+
+		$new_xml = $this->get_unprocessed_xml() . $next_chunk;
+		$breadcrumbs = $this->get_breadcrumbs();
+		$parser_context = $this->get_parser_context();
+
+		$this->reset_state();
+
+		$this->xml = $new_xml;
+		$this->stack_of_open_elements = $breadcrumbs;
+		$this->parser_context         = $parser_context;
+		$this->had_previous_chunks    = true;
```

Comment:
@dmsnell My thinking is the processor has no idea whether the input stream is finished or not. It can make an assumption that an unclosed tag means we're paused at incomplete input, but that input may be in fact complete. The reverse is also problematic – we may have enough input to infer parsing is finished when in fact more input is coming. Perhaps these processors need to be explicitly told "we're still waiting for more data" or "no more input data will come".

> with the HTML API's extend() I've planned on ensuring that we only cut off as much from the front of the document until the first bookmark.

Are there any system-level bookmarks that are implicitly created? As in, is there a chance we'd never forget any bytes because we've seen the `<body>` tag and we'll keep track of it indefinitely?

A memory limit also crossed my mind, as in "never buffer more than 1MB of data", although that seems more complex and maybe not worth it.

Reply:
> Perhaps these processors need to be explicitly told "we're still waiting for more data" or "no more input data will come".

Yes I believe this is going to be the demand. At some point I think we will probably add some method like `$processor->get_incomplete_end_of_document()` but it's not there because I have no idea what that should be right now, or if it's truly necessary.

Only the caller will be able to know if the document was truncated or if more chunks are inbound. This is also true for cases where we have everything in memory, e.g. we got truncated HTML as input and don't know where it came from - "that's it, that's all!"

> Are there any system-level bookmarks that are implicitly created? As in, is there a chance we'd never forget any bytes because we've seen the <body> tag and we'll keep track of it indefinitely?

In the HTML Processor there are for sure, though in the case of the fragment parser, since the context element never exists on the stack of open elements this shouldn't be a problem. We should be able to eject portions of the string that are closed.